File: backend/src/Application/Portraits/OwnPortrait.cs
// Copyright © 2026 Krzysztof Kasprowicz
// Licensed under the GNU Affero General Public License, Version 3. See LICENSE in the project root for license information.
// Project repository: https://github.com/Krzysztof318/MailFathom

using MailFathom.Application.Access;
using MailFathom.Application.StoredFiles;
using MailFathom.Domain.Access;

namespace MailFathom.Application.Portraits;

/// <summary>Reads, replaces, and removes the picture the signed-in person is drawn by.</summary>
/// <remarks>
/// <para>
/// Whose portrait this is comes from the principal rather than from the request, exactly as it does for the user
/// record and for a person's preferences: there is no argument here for another user's identifier, so a request about
/// somebody else is something a caller cannot express rather than something a surface has to refuse.
/// </para>
/// <para>
/// All three acts are admitted under <see cref="MailFathomPermission.MailRead" />, the grant a signed-in person
/// already holds, and none of them adds a name to the published set. The two writes are deliberately not
/// <see cref="MailFathomPermission.MailAccountsWrite" />: that grant decides which mailboxes this deployment connects
/// to, an administrator maintains it for somebody whose accounts are declared in the deployment's own files, and what
/// a person is drawn by must not be decided by a grant over their mail configuration.
/// </para>
/// <para>
/// The picture is a stored file the user record links to. A replacement writes the new file, then links it, then
/// removes the file it displaced, so a failure between any two steps leaves at most a file nothing links to, which
/// <see cref="UnlinkedStoredFileSweep" /> removes — never a record linking to a file that is gone.
/// </para>
/// <para>
/// A read answers nothing where there is no portrait, rather than refusing: a client draws the initials it already has
/// from the person's name, so an absent picture is an ordinary state of the screen rather than an error on it. Octets
/// that are no image kind this build stores are answered the same way, because a file nothing here could have written
/// is a file this surface has nothing to say about.
/// </para>
/// </remarks>
public sealed class OwnPortrait
{
    private readonly AccessAuthorization authorization;
    private readonly IStoredFileStore files;
    private readonly IUserRecordFileLinks links;

    /// <summary>Initializes the use case.</summary>
    /// <param name="authorization">Reports the grant the caller holds and the user it acts for.</param>
    /// <param name="files">Holds the file the portrait is.</param>
    /// <param name="links">Reads and rewrites the link from the user record to that file.</param>
    /// <exception cref="ArgumentNullException">Thrown when any argument is <see langword="null" />.</exception>
    public OwnPortrait(AccessAuthorization authorization, IStoredFileStore files, IUserRecordFileLinks links)
    {
        ArgumentNullException.ThrowIfNull(authorization);
        ArgumentNullException.ThrowIfNull(files);
        ArgumentNullException.ThrowIfNull(links);

        this.authorization = authorization;
        this.files = files;
        this.links = links;
    }

    /// <summary>Reads the picture the signed-in person is drawn by.</summary>
    /// <param name="cancellationToken">Propagates caller cancellation.</param>
    /// <returns>Their portrait, or <see langword="null" /> where this deployment holds none for them.</returns>
    /// <exception cref="PrincipalNotAuthorizedException">Thrown when the caller acts for no user, or its grant omits <see cref="MailFathomPermission.MailRead" />.</exception>
    public async Task<UserPortrait?> ReadAsync(CancellationToken cancellationToken)
    {
        this.authorization.RequirePermission(MailFathomPermission.MailRead);

        var user = this.authorization.RequireUser();

        if (await this.links.FindPortraitAsync(user, cancellationToken) is not { } linked)
        {
            return null;
        }

        return await this.files.ReadAsync(user, linked, cancellationToken) is { } content
            ? UserPortrait.Of(content)
            : null;
    }

    /// <summary>Replaces the picture the signed-in person is drawn by.</summary>
    /// <param name="portrait">The portrait, whose octets are stored as they were supplied.</param>
    /// <param name="cancellationToken">Propagates caller cancellation.</param>
    /// <returns><see langword="true" /> when the write landed, and <see langword="false" /> when this deployment holds no record for the caller.</returns>
    /// <exception cref="ArgumentNullException">Thrown when <paramref name="portrait" /> is <see langword="null" />.</exception>
    /// <exception cref="PrincipalNotAuthorizedException">Thrown when the caller acts for no user, or its grant omits <see cref="MailFathomPermission.MailRead" />.</exception>
    public async Task<bool> ReplaceAsync(UserPortrait portrait, CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(portrait);

        this.authorization.RequirePermission(MailFathomPermission.MailRead);

        var user = this.authorization.RequireUser();

        if (await this.files.WriteAsync(user, portrait.Type.MediaType, portrait.Content, cancellationToken) is not { } written)
        {
            return false;
        }

        var relinked = await this.links.RelinkPortraitAsync(user, written, cancellationToken);

        if (!relinked.UserHeld)
        {
            return false;
        }

        if (relinked.Replaced is { } displaced && displaced != written)
        {
            await this.files.RemoveAsync(user, displaced, cancellationToken);
        }

        return true;
    }

    /// <summary>Removes the picture the signed-in person is drawn by.</summary>
    /// <param name="cancellationToken">Propagates caller cancellation.</param>
    /// <exception cref="PrincipalNotAuthorizedException">Thrown when the caller acts for no user, or its grant omits <see cref="MailFathomPermission.MailRead" />.</exception>
    /// <remarks>
    /// Nothing else about the person is touched, and the removal is silent about whether there was one to remove. The
    /// link goes before the file does, for the reason a replacement orders its steps.
    /// </remarks>
    public async Task RemoveAsync(CancellationToken cancellationToken)
    {
        this.authorization.RequirePermission(MailFathomPermission.MailRead);

        var user = this.authorization.RequireUser();
        var relinked = await this.links.RelinkPortraitAsync(user, null, cancellationToken);

        if (relinked.Replaced is { } displaced)
        {
            await this.files.RemoveAsync(user, displaced, cancellationToken);
        }
    }
}

File: frontend/src/Client.Backend/src/ownPortrait.ts
// Copyright © 2026 Krzysztof Kasprowicz
// Licensed under the GNU Affero General Public License, Version 3. See LICENSE in the project root for license information.
// Project repository: https://github.com/Krzysztof318/MailFathom

import type { ClientFailureReason } from './failure';
import { headersFor, routeFor, type ClientSession } from './session';
import { reported } from './telemetry';
import type { ClientRequest } from './transport';

// The picture the signed-in person is drawn by. Like a message's attachment, and for the same reason, this module
// composes the requests and hands them to an adapter to send: what a read answers with is octets and what a
// replacement carries is a file, and a `MailFathomTransport` speaks in text — so putting these on the wire is the
// application's, exactly as calling `fetch` at all is.
//
// What stays here is everything the boundary owns: the route, the credential, the kinds this surface accepts, the
// bound an upload is refused past, the header the answer is asked in, and the record kept of each request. No screen
// and no adapter above learns a path or a header name, and none of them decides what these are called in a trace.

/** The route the acting person's portrait is read, replaced, and removed at, relative to the client prefix. */
export const ownPortraitRoute = '/portrait';

/**
 * The largest portrait this surface accepts, in octets.
 *
 * It is the deployment's own bound restated rather than a second opinion: an upload past it is refused there before a
 * handler is entered, and a client that knew nothing of it would let somebody wait for a megabyte to travel in order
 * to be told no. It is also the bound a read is taken under, because a portrait larger than the one that could have
 * been stored is not one this deployment wrote.
 */
export const largestPortraitOctets = 1_048_576;

/** The kinds of picture this surface stores, which is the whole set rather than a preference among more. */
export const portraitImageTypes = ['image/jpeg', 'image/png'] as const;

/** One kind of picture a portrait may be. */
export type PortraitImageType = (typeof portraitImageTypes)[number];

/** Whether a kind a browser reported for a chosen file is one this surface stores. */
export function isPortraitImageType(value: string): value is PortraitImageType {
    return portraitImageTypes.includes(value as PortraitImageType);
}

/**
 * Composes the request that reads the picture the signed-in person is drawn by.
 *
 * The answer is octets rather than JSON, so the request asks for the two kinds this surface serves; the credential is
 * unchanged, being the access control the route applies.
 */
export function readOwnPortraitRequest(session: ClientSession): ClientRequest {
    return {
        method: 'GET',
        path: routeFor(session, ownPortraitRoute),
        headers: { ...headersFor(session), Accept: portraitImageTypes.join(', ') },
        longestAnswer: largestPortraitOctets,
    };
}

/**
 * Composes the request that replaces the picture, under the kind the chosen file was found to be.
 *
 * The octets are not here: a file is a browser value this package may not name, so the adapter that puts this request
 * on the wire carries them. What this states is the kind they will be sent under, which the deployment reads the
 * signature against rather than trusting.
 */
export function replaceOwnPortraitRequest(session: ClientSession, type: PortraitImageType): ClientRequest {
    return {
        method: 'POST',
        path: routeFor(session, ownPortraitRoute),
        headers: { ...headersFor(session), 'Content-Type': type },
    };
}

/** Composes the request that removes the picture, leaving everything else about the person as it was. */
export function removeOwnPortraitRequest(session: ClientSession): ClientRequest {
    return {
        method: 'DELETE',
        path: routeFor(session, ownPortraitRoute),
        headers: headersFor(session),
    };
}

/**
 * Reads the picture through an adapter, and reports the request the way every other one on this surface is.
 *
 * The three operations below each compose inside their own span for the reason `readMailAttachment` gives: the trace
 * context is written from whatever span is active while the request is composed, and what a request is named in a
 * trace is this package's decision rather than the caller's.
 *
 * @param deliver Puts the composed request on the wire and answers whatever the application makes of it.
 * @param failureOf Which failure that answer amounts to, or `null` where the client got an answer it acts on — a
 * person with no portrait stored among them, that being a state the screen draws rather than one it reports.
 */
export function readOwnPortrait<TOutcome>(
    session: ClientSession,
    deliver: (request: ClientRequest) => Promise<TOutcome>,
    failureOf: (outcome: TOutcome) => ClientFailureReason | null,
): Promise<TOutcome> {
    return reported(`GET ${ownPortraitRoute}`, () => deliver(readOwnPortraitRequest(session)), failureOf);
}

/** Replaces the picture through an adapter, which is what carries the octets, and reports the request. */
export function replaceOwnPortrait<TOutcome>(
    session: ClientSession,
    type: PortraitImageType,
    deliver: (request: ClientRequest) => Promise<TOutcome>,
    failureOf: (outcome: TOutcome) => ClientFailureReason | null,
): Promise<TOutcome> {
    return reported(`POST ${ownPortraitRoute}`, () => deliver(replaceOwnPortraitRequest(session, type)), failureOf);
}

/** Removes the picture through an adapter, and reports the request. */
export function removeOwnPortrait<TOutcome>(
    session: ClientSession,
    deliver: (request: ClientRequest) => Promise<TOutcome>,
    failureOf: (outcome: TOutcome) => ClientFailureReason | null,
): Promise<TOutcome> {
    return reported(`DELETE ${ownPortraitRoute}`, () => deliver(removeOwnPortraitRequest(session)), failureOf);
}
